In [151]:

import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier

In [152]:
df = pd.read_csv('weather.csv', sep=';')
df

Month Location  MinTemp  MaxTemp  Rainfall  WindGustSpeed  \
0          1   Albury     20.4     37.6       0.0             54   
1          1   Albury     20.9     33.6       0.4             50   
2          1   Albury     18.4     23.1       2.2             48   
3          1   Albury     17.3     23.7      15.6             39   
4          1   Albury     15.5     22.9       6.8             31   
...      ...      ...      ...      ...       ...            ...   
14463     12    Uluru     22.1     35.8      63.8             43   
14464     12    Uluru     22.6     36.8       0.0             50   
14465     12    Uluru     23.2     38.0       0.0             33   
14466     12    Uluru     19.7     37.0       0.0             37   
14467     12    Uluru     23.7     33.0       0.0             46   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0                 0             7           46           17       1013.4   
1                 9            17           54           30       1011.1   
2                11            39           62           67       1014.0   
3                 9            17           74           65       1017.9   
4                 6             9           92           63       1016.3   
...             ...           ...          ...          ...          ...   
14463            17            31           77           41        997.8   
14464            30            15           63           39       1000.0   
14465            17            17           25           14       1004.4   
14466            22             6           30           23       1004.6   
14467            24            17           40           38       1005.1   

       Pressure3pm  Cloud3pm  Temp9am  Temp3pm RainToday RainTomorrow  
0           1009.2         3     26.1     36.7        No           No  
1           1008.4         8     24.8     31.7        No          Yes  
2           1014.8         8     21.8     19.5       Yes          Yes  
3           1016.5         8     19.2     21.6       Yes          Yes  
4           1013.9         8     17.2     22.2       Yes           No  
...            ...       ...      ...      ...       ...          ...  
14463        995.2         2     26.0     33.5       Yes           No  
14464        998.8        -1     29.7     34.0        No           No  
14465       1001.0        -1     28.7     36.4        No           No  
14466       1000.9        -1     28.2     35.1        No           No  
14467       1002.7        -1     30.1     31.5        No           No  

[14468 rows x 17 columns]

# 1
Si vuole predire la possibilità di avere pioggia il giorno successivo (RainTomorrow è la classe
da predire).
Trasformare i valori degli attributi RainToday e RainTomorrow da No a 0, e da Yes a 1.
Creare un nuovo dataset chiamato reduce con le istanze del dataset per le quali c’è un valore di
Cloud3pm maggiore o uguale a 0.
Dividere il dataset in modo che 4/5 degli elementi siano contenuti in un nuovo dataset “train” e
1/5 in un dataset “test”.
Valutare l’accuracy ottenuta con il classificatore Logistic Regression e il Decision Tree
Il valore di accuratezza ottenuto è pari a _____________ peculiarità? (punti 4)
La confusion matrix evidenzia delle

In [153]:
mapping = {
    'No': 0,
    'Yes' : 1,
}
df['RainTomorrow'] = df['RainTomorrow'].map(mapping)
df['RainToday'] = df['RainToday'].map(mapping)
df

Month Location  MinTemp  MaxTemp  Rainfall  WindGustSpeed  \
0          1   Albury     20.4     37.6       0.0             54   
1          1   Albury     20.9     33.6       0.4             50   
2          1   Albury     18.4     23.1       2.2             48   
3          1   Albury     17.3     23.7      15.6             39   
4          1   Albury     15.5     22.9       6.8             31   
...      ...      ...      ...      ...       ...            ...   
14463     12    Uluru     22.1     35.8      63.8             43   
14464     12    Uluru     22.6     36.8       0.0             50   
14465     12    Uluru     23.2     38.0       0.0             33   
14466     12    Uluru     19.7     37.0       0.0             37   
14467     12    Uluru     23.7     33.0       0.0             46   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0                 0             7           46           17       1013.4   
1                 9            17           54           30       1011.1   
2                11            39           62           67       1014.0   
3                 9            17           74           65       1017.9   
4                 6             9           92           63       1016.3   
...             ...           ...          ...          ...          ...   
14463            17            31           77           41        997.8   
14464            30            15           63           39       1000.0   
14465            17            17           25           14       1004.4   
14466            22             6           30           23       1004.6   
14467            24            17           40           38       1005.1   

       Pressure3pm  Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  
0           1009.2         3     26.1     36.7          0             0  
1           1008.4         8     24.8     31.7          0             1  
2           1014.8         8     21.8     19.5          1             1  
3           1016.5         8     19.2     21.6          1             1  
4           1013.9         8     17.2     22.2          1             0  
...            ...       ...      ...      ...        ...           ...  
14463        995.2         2     26.0     33.5          1             0  
14464        998.8        -1     29.7     34.0          0             0  
14465       1001.0        -1     28.7     36.4          0             0  
14466       1000.9        -1     28.2     35.1          0             0  
14467       1002.7        -1     30.1     31.5          0             0  

[14468 rows x 17 columns]

In [154]:
reduce = df[df['Cloud3pm'] >= 0]
X = reduce.drop('RainTomorrow', axis=1)
y = reduce['RainTomorrow']

le = LabelEncoder()
X['Location'] = le.fit_transform(X['Location'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

lr = LogisticRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_train)
acc_lr_train = accuracy_score(y_pred, y_train)
y_pred = lr.predict(X_test)
acc_lr_test = accuracy_score(y_pred, y_test)

print('---LOGISTIC REGRESSION---')
print(f'Accuracy train: {acc_lr_train}, Accuracy test: {acc_lr_test}')
print(confusion_matrix(y_test, y_pred))

---LOGISTIC REGRESSION---
Accuracy train: 0.8112359550561797, Accuracy test: 0.8082684242061114
[[1100   90]
 [ 230  249]]


In [155]:
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)
y_pred = dt.predict(X_train)
acc_dt_train = accuracy_score(y_pred, y_train)
y_pred = dt.predict(X_test)
acc_dt_test = accuracy_score(y_pred, y_test)
print('---Decision Tree Classifier---')
print(f'Accuracy train: {acc_dt_train}, Accuracy test: {acc_dt_test}')
print(confusion_matrix(y_test, y_pred))

---Decision Tree Classifier---
Accuracy train: 1.0, Accuracy test: 0.7417615338526063
[[980 210]
 [221 258]]


# 2
Che valore di accuratezza si ottiene con un 5 Fold cross validation e il classificatore basato su
Decision Tree ________________ e quello basato su Logistic Regression ________________
Il valore di accuratezza maggiormente rappresentativo è quello che si ottiene con questa tecnica
o con quella attuata in precedenza? (punti 2)

In [156]:
import warnings
warnings.filterwarnings('ignore')
cross = cross_val_score(dt, X, y, cv=5, n_jobs=-1, scoring='accuracy')
print('---Cross Validation DECISION TREE---')
print(cross)
print(cross.mean())
cross = cross_val_score(lr, X, y, cv=5, n_jobs=-1, scoring='accuracy')
print('---Cross Validation LOGISTIC REGRESSION---')
print(cross)
print(cross.mean())


---Cross Validation DECISION TREE---
[0.72139005 0.72079089 0.74355902 0.73876573 0.71223022]
0.727347181571699
---Cross Validation LOGISTIC REGRESSION---
[0.80886759 0.78789694 0.81366087 0.81905333 0.78956835]
0.8038094150204103


Il valore ottenuto con la Cross Validation è maggiormente rappresentativo, perché la divisione singola è sensibile alla casualità. La Cross Validation, invece, media i risultati su 5 diverse partizioni dell'intero dataset, riducendo la varianza e fornendo una stima più affidabile di come il modello si comporterà su dati futuri mai visti.

# 3
Si introduca un attributo che sostituisca per ogni rilevazione i due valori di temperatura
“MinTemp, MaxTemp” con il valore medio delle registrazioni. Si faccia lo stesso con vento
(WindSpeed9am, WindSpeed3pm). Che valore di accuratezza si ottiene? ________________
(punti 4)

In [157]:
df['Temp'] = (df['Temp3pm'] + df['Temp9am']) /2
df['Wind'] = (df['WindSpeed3pm'] + df['WindSpeed9am']) /2
X = df.drop(['WindSpeed3pm', 'WindSpeed9am', 'Temp3pm', 'Temp9am', 'RainTomorrow'], axis=1)
y = df['RainTomorrow']
X['Location'] = le.fit_transform(X['Location'])
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [158]:

lr = LogisticRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_train)
acc_lr_fix_train = accuracy_score(y_pred, y_train)
y_pred = lr.predict(X_test)
acc_lr_fix_test = accuracy_score(y_pred, y_test)

print('---LOGISTIC REGRESSION---')
print(f'Before: Accuracy train: {acc_lr_train}, Accuracy test: {acc_lr_test}')
print(f'FIXED: Accuracy train: {acc_lr_fix_train}, Accuracy test: {acc_lr_fix_test}')
print(confusion_matrix(y_test, y_pred))

---LOGISTIC REGRESSION---
Before: Accuracy train: 0.8112359550561797, Accuracy test: 0.8082684242061114
FIXED: Accuracy train: 0.8389493692759634, Accuracy test: 0.834139599170698
[[2110  128]
 [ 352  304]]


In [159]:
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)
y_pred = dt.predict(X_train)
acc_dt_fix_train = accuracy_score(y_pred, y_train)
y_pred = dt.predict(X_test)
acc_dt_fix_test = accuracy_score(y_pred, y_test)
print('---Decision Tree Classifier---')
print(f'Before: Accuracy train: {acc_dt_train}, Accuracy test: {acc_dt_test}')
print(f'FIXED: Accuracy train: {acc_dt_fix_train}, Accuracy test: {acc_dt_fix_test}')
print(confusion_matrix(y_test, y_pred))

---Decision Tree Classifier---
Before: Accuracy train: 1.0, Accuracy test: 0.7417615338526063
FIXED: Accuracy train: 1.0, Accuracy test: 0.7736696613683484
[[1898  340]
 [ 315  341]]


# 4
Si riparta dal dataset originario e si considerino due nuovi dataset ottenuti rimuovendo dal
dataset di partenza gli elementi con Cloud3pm minore di 0. Il dataset con i valori negativi si
chiamerà cloudP, l’altro cloudT.
Si alleni un regressore su cloudT per predire i valori di Cloud3pm. Si usi il modello per sostituire
in cloudP il valore predetto per Cloud3pm. (punti 4)

In [160]:
df = pd.read_csv('weather.csv', sep=';')
df['Location'] = le.fit_transform(df['Location'])
df['RainTomorrow'] = df['RainTomorrow'].map(mapping)
df['RainToday'] = df['RainToday'].map(mapping)
df

Month  Location  MinTemp  MaxTemp  Rainfall  WindGustSpeed  \
0          1         1     20.4     37.6       0.0             54   
1          1         1     20.9     33.6       0.4             50   
2          1         1     18.4     23.1       2.2             48   
3          1         1     17.3     23.7      15.6             39   
4          1         1     15.5     22.9       6.8             31   
...      ...       ...      ...      ...       ...            ...   
14463     12        34     22.1     35.8      63.8             43   
14464     12        34     22.6     36.8       0.0             50   
14465     12        34     23.2     38.0       0.0             33   
14466     12        34     19.7     37.0       0.0             37   
14467     12        34     23.7     33.0       0.0             46   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0                 0             7           46           17       1013.4   
1                 9            17           54           30       1011.1   
2                11            39           62           67       1014.0   
3                 9            17           74           65       1017.9   
4                 6             9           92           63       1016.3   
...             ...           ...          ...          ...          ...   
14463            17            31           77           41        997.8   
14464            30            15           63           39       1000.0   
14465            17            17           25           14       1004.4   
14466            22             6           30           23       1004.6   
14467            24            17           40           38       1005.1   

       Pressure3pm  Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  
0           1009.2         3     26.1     36.7          0             0  
1           1008.4         8     24.8     31.7          0             1  
2           1014.8         8     21.8     19.5          1             1  
3           1016.5         8     19.2     21.6          1             1  
4           1013.9         8     17.2     22.2          1             0  
...            ...       ...      ...      ...        ...           ...  
14463        995.2         2     26.0     33.5          1             0  
14464        998.8        -1     29.7     34.0          0             0  
14465       1001.0        -1     28.7     36.4          0             0  
14466       1000.9        -1     28.2     35.1          0             0  
14467       1002.7        -1     30.1     31.5          0             0  

[14468 rows x 17 columns]

In [161]:
cloudP = df[df['Cloud3pm'] < 0]
cloudT = df[df['Cloud3pm'] >= 0]

X_cloudP = cloudP.drop('Cloud3pm', axis=1)
y_cloudP = cloudP['Cloud3pm']

X_cloudT = cloudT.drop('Cloud3pm', axis=1)
y_cloudT = cloudT['Cloud3pm']

In [162]:
lr.fit(X_cloudT, y_cloudT)
y_pred = lr.predict(X_cloudP)
print(accuracy_score(y_cloudP, y_pred))
print(y_pred)
cloudP['Cloud3pm'] = y_pred
print(cloudP[cloudP['Cloud3pm'] > 1])
cloudP

0.0
[1 1 1 ... 1 1 1]
       Month  Location  MinTemp  MaxTemp  Rainfall  WindGustSpeed  \
124        5         1     11.0     17.5       2.4             37   
133        5         1     13.6     20.2       0.2             24   
140        5         1     12.1     18.5       0.0             35   
141        5         1      4.7     17.9       0.0             13   
142        5         1      5.3     21.8       0.0             20   
...      ...       ...      ...      ...       ...            ...   
14242      5        34     13.5     26.4       0.0             33   
14268      5        34     13.7     23.0       2.4             22   
14279      6        34     10.2     15.6      42.0             43   
14285      7        34      9.4     16.4       4.8             39   
14293      7        34     12.4     18.8      13.6             43   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
124               9             9           73           39       1017.2 

Month  Location  MinTemp  MaxTemp  Rainfall  WindGustSpeed  \
9          1         1     17.8     37.0       0.0             41   
11         1         1     20.6     38.9       0.0             56   
12         1         1     20.2     43.0       0.0             72   
14         1         1     10.4     25.2       0.0             46   
15         1         1     10.7     29.4       0.0             35   
...      ...       ...      ...      ...       ...            ...   
14457     12        34     26.1     38.2       0.0             50   
14464     12        34     22.6     36.8       0.0             50   
14465     12        34     23.2     38.0       0.0             33   
14466     12        34     19.7     37.0       0.0             37   
14467     12        34     23.7     33.0       0.0             46   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
9                 9            17           54           17       1016.6   
11                6            24           40           18       1010.8   
12               15            24           51           16       1013.8   
14               19            20           41           31       1026.1   
15               13             9           45           19       1027.4   
...             ...           ...          ...          ...          ...   
14457            26            26           32           19       1010.3   
14464            30            15           63           39       1000.0   
14465            17            17           25           14       1004.4   
14466            22             6           30           23       1004.6   
14467            24            17           40           38       1005.1   

       Pressure3pm  Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  
9           1012.6         1     26.2     36.2          0             0  
11          1009.8         1     28.1     37.6          0             0  
12          1007.5         1     27.9     42.3          0             0  
14          1023.8         1     15.9     23.9          0             0  
15          1022.9         1     17.8     27.3          0             0  
...            ...       ...      ...      ...        ...           ...  
14457       1006.4         1     29.1     37.1          0             0  
14464        998.8         1     29.7     34.0          0             0  
14465       1001.0         1     28.7     36.4          0             0  
14466       1000.9         1     28.2     35.1          0             0  
14467       1002.7         1     30.1     31.5          0             0  

[6124 rows x 17 columns]

# 5
Si consideri il dataset ottenuto concatenando cloudP e cloudT in un unico dataset e si
confronti l’accuratezza che si ottiene con un 10 Fold cross validation e il classificatore basato su
Decision Tree e quello basato su Logistic Regression con quella ottenuta al punto 3. (punti 3)

In [172]:
df_conc = pd.concat([cloudP, cloudT])

X = df_conc.drop('RainTomorrow', axis=1)
y = df_conc['RainTomorrow']

cross_conc = cross_val_score(dt, X, y, cv=10, n_jobs=-1, scoring='accuracy')
print('---Cross Validation DECISION TREE---')
print(cross)
print(cross_conc)
print(f'PRIMA: {cross.mean()}')
print(f'CONCATENATO: {cross_conc.mean()}')

cross_conc = cross_val_score(lr, X, y, cv=10, n_jobs=-1, scoring='accuracy')
print('---Cross Validation LOGISTIC REGRESSION---')
print(cross)
print(cross_conc)
print(f'PRIMA: {cross.mean()}')
print(f'CONCATENATO: {cross_conc.mean()}')


---Cross Validation DECISION TREE---
[0.80886759 0.78789694 0.81366087 0.81905333 0.78956835]
[0.72978576 0.73807878 0.73047685 0.6129924  0.74775397 0.5805114
 0.67795439 0.74775397 0.74204703 0.73236515]
PRIMA: 0.8038094150204103
CONCATENATO: 0.7039719704334144
---Cross Validation LOGISTIC REGRESSION---
[0.80886759 0.78789694 0.81366087 0.81905333 0.78956835]
[0.77816171 0.85072564 0.82791983 0.861783   0.83621285 0.84727021
 0.77608846 0.83621285 0.8264177  0.81258645]
PRIMA: 0.8038094150204103
CONCATENATO: 0.8253378717449467


Logistic Regression (Migliorata): L'operazione di recupero dei dati mancanti (prevedendo Cloud3pm con la Regressione Lineare) è stata molto efficace. L'accuracy è salita a 0.825 (82.5%) nel dataset concatenato, superando il valore base (~80.8%) e quello ottenuto con la media. Questo indica che la nuvolosità è una variabile predittiva forte e averla "ricostruita" ha aiutato il modello.

Decision Tree (Peggiorata): Al contrario, il Decision Tree ha performato peggio sul dataset concatenato, scendendo a 0.701 (70%). Probabilmente il modello è troppo sensibile al "rumore" introdotto dai valori predetti artificialmente e non riesce a sfruttarli correttamente.

# 6
Utilizzare un algoritmo di regressione da applicarsi al dataset del punto 1 per effettuare la
predizione. Arrotondare i valori predetti all’intero. Confrontare i risultati ottenuti con quelli
ottenuti nei punti precedenti (punti 3).

In [171]:
X = reduce.drop('RainTomorrow', axis=1)
y = reduce['RainTomorrow']

le = LabelEncoder()
X['Location'] = le.fit_transform(X['Location'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

y_pred_lin_raw = lin_reg.predict(X_test)
print(y_pred_lin_raw)
y_pred_lin_rounded = np.round(y_pred_lin_raw)
acc_lin = accuracy_score(y_test, y_pred_lin_rounded)

print("--- Linear Regression (usata come classificatore) ---")
print(f"Accuracy: {acc_lin:.4f}")
confronto = pd.DataFrame({
    'Reale': y_test,
    'Predizione_Grezza': y_pred_lin_raw,
    'Predizione_Arrotondata': y_pred_lin_rounded
})
confronto

[0.03102402 0.10730006 0.75005995 ... 0.66018275 0.26257848 0.69728974]
--- Linear Regression (usata come classificatore) ---
Accuracy: 0.8053


Reale  Predizione_Grezza  Predizione_Arrotondata
11667      0           0.031024                     0.0
8714       0           0.107300                     0.0
6959       1           0.750060                     1.0
2175       1           0.324405                     0.0
8877       0           0.512374                     1.0
...      ...                ...                     ...
2570       0           0.033184                     0.0
12154      0           0.208841                     0.0
13817      1           0.660183                     1.0
3780       1           0.262578                     0.0
9029       1           0.697290                     1.0

[1669 rows x 3 columns]